### PR TITLE
Run flynt to convert %-formatting to f-strings

### PR DESCRIPTION
**Pull request by eigenfoo**
~[WIP, will request review when ready]~

A follow-up to #139 - I simply used the [`flynt` CLI tool](https://github.com/ikamensh/flynt) to automatically convert %-formatted strings to f-strings. However, not all formatted strings have been converted to f-strings - some strings are tricky to convert, or `flynt` simply fails on them - in which case I've just left the string as-is.

It's important to note that %-formatting are different from f-strings [in meaningful ways](https://github.com/ikamensh/flynt#dangers-of-conversion), so it's important to check the tests.

---

For posterity, here's the output of `flynt`:

```
Running flynt v.0.56

Flynt run has finished. Stats:

Execution time:                            19.170s
Files modified:                            125
Character count reduction:                 2509 (0.03%)

Per expression type:
Old style (`%`) expressions attempted:     846/1483 (57.0%)
`.format(...)` calls attempted:            286/291 (98.3%)
No concatenations attempted.
F-string expressions created:              799
Out of all attempted transforms, 321 resulted in errors.
To find out specific error messages, use --verbose flag.

_-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_.
Please run your tests before committing. Did flynt get a perfect conversion? give it a star at:
~ https://github.com/ikamensh/flynt ~
Thank you for using flynt. Upgrade more projects and recommend it to your colleagues!
```

### CODE REVIEW

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Changes needed:
- [x] I believe that `theano/_version.py` is generated automatically by `versioneer`; if so, we're better off regenerating that file and/or leaving it untouched.
- [x] There are _a lot_ of f-strings containing superfluous uses of `str`, `dict`, and `locals`.  

It would seem as though `flynt` isn't aware of these redundancies.  If you're interested in writing your own refactoring code that _is_ aware of such things, check out [LibCST](https://github.com/Instagram/LibCST)!

**Comment by eigenfoo** [comment]
@brandonwillard not to dismiss your suggestion, but I just didn't feel like adding to my cognitive load by learning a new tool - I just `rg`ed the diff to find offending instances of `locals()`, `dict`s and `str`s and manually fixed them with some help from Vim macros.

**Comment by twiecki** [comment]
@eigenfoo seems like tests still fail.

**Comment by eigenfoo** [comment]
The only test that's failing is the patch coverage, which seems to be failing for all PRs that don't test 100% of their changes. Since this PR doesn't change any actual code behavior, I think that's okay.

**Comment by twiecki** [comment]
Gotcha, cool.

**Comment by brandonwillard** [comment]
> ...I just didn't feel like adding to my cognitive load by learning a new tool...

Sorry, I wasn't actually suggesting that you use LibCST for the work in this PR.  I was pointing you in the direction of a library that's quite useful for creating your own fully-automated Python refactorings.  It's more of a long-term suggestion, since I strongly believe that Python needs more/better refactoring tools and libraries and I think that's more likely to happen if/when libraries like LibCST are given more attention.

File: theano/compile/profiling.py [hunk 1]
```diff
@@ -1452,7 +1452,7 @@ def print_stats(stats1, stats2):
                 size = "%10s" % "Unknown"
```
Comment (brandonwillard) [inline theano/compile/profiling.py:1452]: I wonder why it didn't pick up this one.

File: theano/gof/cc.py [hunk 2]
```diff
@@ -888,7 +885,7 @@ def code_gen(self):
             try:
                 behavior = op.c_code(node, name, isyms, osyms, sub)
             except utils.MethodNotDefined:
-                raise NotImplementedError("%s cannot produce C code" % op)
+                raise NotImplementedError(f"{op} cannot produce C code")
             assert isinstance(behavior, str), (
                 str(node.op) + " didn't return a string for c_code"
```
Comment (brandonwillard) [inline theano/gof/cc.py:890]: This looks like a good place for an f-string.

File: theano/gof/compiledir.py [hunk 3]
```diff
@@ -173,7 +173,7 @@ def print_compiledir_content():
             except AttributeError:
                 _logger.error("Could not read key file '%s'.", filename)
 
-    print_title("Theano cache: %s" % compiledir, overline="=", underline="=")
+    print_title(f"Theano cache: {compiledir}", overline="=", underline="=")
     print()
 
     print_title("List of %d compiled individual ops" % len(table), underline="+")
```
Comment (brandonwillard) [inline theano/gof/compiledir.py:179]: This could also be an f-string.

File: theano/gpuarray/sort.py [hunk 4]
```diff
@@ -224,8 +223,8 @@ def c_code(self, node, nodename, inps, outs, sub):
         dims = "".join("dims[%d], " % i for i in reordered_axes[1:])
```
Comment (brandonwillard) [inline theano/gpuarray/sort.py:223]: This could be an f-string, too.

File: theano/gpuarray/sort.py [hunk 5]
```diff
@@ -240,8 +239,8 @@ def c_code(self, node, nodename, inps, outs, sub):
             def_dvstrides = params_dv = ""
 
         if self.return_indices:
-            def_distrides = "const ssize_t *distrides = PyGpuArray_STRIDES(%s)" % yi
-            params_di = "{}->ga.data, {}->ga.offset,\n".format(yi, yi)
+            def_distrides = f"const ssize_t *distrides = PyGpuArray_STRIDES({yi})"
+            params_di = f"{yi}->ga.data, {yi}->ga.offset,\n"
             params_di += "".join("distrides[%d], " % i for i in reordered_axes)
```
Comment (brandonwillard) [inline theano/gpuarray/sort.py:244]: Same here.

File: theano/gpuarray/type.py [hunk 6]
```diff
@@ -235,9 +233,7 @@ def __repr__(self):
                 bcast = str(b)
             else:
                 bcast = "%iD" % len(b)
```
Comment (brandonwillard) [inline theano/gpuarray/type.py:235]: This could also be an f-string.

File: theano/gradient.py [hunk 7]
```diff
@@ -838,9 +837,9 @@ def _node_to_pattern(node):
         for ii, output_pattern in enumerate(connection_pattern):
             if not isinstance(output_pattern, list):
                 raise TypeError(
-                    "%s.connection_pattern should return" % node.op
+                    f"{node.op}.connection_pattern should return"
                     + " a list of lists, but element %d" % ii
```
Comment (brandonwillard) [inline theano/gradient.py:841]: Another f-string candidate.

File: theano/misc/check_blas.py [hunk 8]
```diff
@@ -293,7 +293,7 @@ def test():
         print("(%d, %d) and (%d, %d)." % (M, N, N, K))
```
Comment (brandonwillard) [inline theano/misc/check_blas.py:293]: This could also be an f-string.

File: theano/misc/hooks/check_whitespace.py [hunk 9]
```diff
@@ -45,9 +45,9 @@ def get_parse_error(code):
     try:
         tabnanny.process_tokens(tokenize.generate_tokens(code_buffer.readline))
     except tokenize.TokenError as err:
-        return "Could not parse code: %s" % err
+        return f"Could not parse code: {err}"
     except IndentationError as err:
-        return "Indentation error: %s" % err
+        return f"Indentation error: {err}"
     except tabnanny.NannyNag as err:
         return "Ambiguous tab at line %d; line is '%s'." % (
```
Comment (brandonwillard) [inline theano/misc/hooks/check_whitespace.py:52]: Another f-string candidate.

File: theano/printing.py [hunk 10]
```diff
@@ -519,7 +519,7 @@ def process(self, output, pstate):
                 " not the result of an operation" % self.function
```
Comment (brandonwillard) [inline theano/printing.py:519]: Another f-string candidate.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Almost there!

It looks like Flynt is restrictively shy about
-  concatenated strings (e.g. `"blah %s " "blah" % "blah"`) and
- interpolations that specify integer presentations (e.g. `"%d"`).

The former can be handled by surrounding the concatenated strings with parentheses and the latter by casting to `int` in the f-string expression.

Otherwise, if I recall, there are some places where the old `%` interpolation is used to create string templates (e.g. `st = "blah %d"; st % 5`).  I don't know if f-strings can be used in that way, so we might need to convert those to `str.format` or `string.Template` forms.  We can create separate issue for that, though.

**Comment by eigenfoo** [comment]
I've removed most instances of `%d` - I've replaced them with `f"...{int(x)}..."` in all places, unless it was obvious that it would be an integer instead of a float (e.g. `f"...{len(x)}..."`). I left them alone when they were in templated strings, which should probably be dealt with in a separate issue/PR.

**Comment by eigenfoo** [comment]
Closed in favor of https://github.com/pymc-devs/Theano-PyMC/pull/162 - we now fork and branch, instead of working off of branches in this repo.

**Pull request closed without merging**